cmd/contour/contour.go

```diff
@@ -126,7 +135,61 @@ func main() {
 			log.WithError(err).Fatal("unable to initialize Server dependencies required to start Contour")
 		}
 
-		if err := serve.doServe(); err != nil {
+		var contourConfiguration contour_api_v1alpha1.ContourConfigurationSpec
+
+		// Get the ContourConfiguration CRD if specified
+		if len(serveCtx.contourConfigurationName) > 0 {
+			// Determine the name/namespace of the configuration resource utilizing the environment
+			// variable "CONTOUR_NAMESPACE" which should exist on the Contour deployment.
+			//
+			// If the env variable is not present, it will default to "projectcontour".
+			contourNamespace, found := os.LookupEnv("CONTOUR_NAMESPACE")
+			if !found {
+				contourNamespace = "projectcontour"
+			}
+
+			namespacedName := types.NamespacedName{Name: serveCtx.contourConfigurationName, Namespace: contourNamespace}
+			client := serve.clients.DynamicClient().Resource(contour_api_v1alpha1.ContourConfigurationGVR).Namespace(namespacedName.Namespace)
+
+			// ensure the specified ContourConfiguration exists
+			res, err := client.Get(context.Background(), namespacedName.Name, metav1.GetOptions{})
+			if err != nil {
+				log.WithError(err).Fatalf("error getting contour configuration %s", namespacedName)
+			}
+
+			var contourConfig contour_api_v1alpha1.ContourConfiguration
+			if err := runtime.DefaultUnstructuredConverter.FromUnstructured(res.Object, &contourConfig); err != nil {
+				log.WithError(err).Fatalf("error converting contour configuration %s", namespacedName)
+			}
+
+			// Set valid status on the ContourConfiguration object
+			contourConfig.Status = contour_api_v1alpha1.ContourConfigurationStatus{
+				Conditions: []contour_api_v1.DetailedCondition{{
+					Condition: contour_api_v1.Condition{
+						Type:               "Valid",
+						Status:             "True",
+						ObservedGeneration: 0,
+						LastTransitionTime: metav1.NewTime(time.Now()),
+						Reason:             "ValidContourConfiguration",
+						Message:            "Valid ContourConfiguration",
+					},
+					Errors:   nil,
+					Warnings: nil,
+				}},
+			}
+
+			if err := serve.sh.SetContourConfigurationStatus(&contourConfig); err != nil {
```

Comment:
This new method doesn't take into account leader election, so all Contour replicas are going to try to write status which could be problematic. Guessing there's an order of operations problem preventing using the existing `serve.sh.Writer().Send()` method since the StatusUpdateHandler hasn't been started yet?

Reply:
Yeah that's a good point. I started down that road but then thought getting to the group.Start() bit was overkill since this will just write then quit. Let me see about doing another refactor to things to allow for this better.